lib/api/CDataFrameAnalyzer.cc

```diff
@@ -247,12 +254,32 @@ void CDataFrameAnalyzer::addRowToDataFrame(const TStrVec& fieldValues) {
     });
 }
 
-void CDataFrameAnalyzer::writeResultsOf(const CDataFrameAnalysisRunner& analysis) const {
-    // TODO Revisit this can probably be core::CRapidJsonLineWriter.
-    // TODO This should probably be a member variable so it is only created once.
-    auto outStream = m_OutStreamSupplier();
-    core::CRapidJsonConcurrentLineWriter outputWriter{*outStream};
+void CDataFrameAnalyzer::reportProgress(const CDataFrameAnalysisRunner& analysis,
+                                        core::CRapidJsonConcurrentLineWriter& writer) const {
+    double progress{0.0};
+    while (analysis.finished() == false) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        double latestProgress{analysis.progress()};
+        if (latestProgress > progress) {
+            progress = latestProgress;
+            this->writeProgress(progress, writer);
+        }
+    }
+    if (progress < 1.0) {
+        this->writeProgress(1.0, writer);
+    }
+}
+
+void CDataFrameAnalyzer::writeProgress(double progress,
+                                       core::CRapidJsonConcurrentLineWriter& writer) const {
+    writer.StartObject();
+    writer.Key(PROGRESS);
+    writer.Double(progress);
```

Comment:
Maybe round to nearest 0.01? High precision isn't needed here and tends to imply we're measuring more accurately than we are.

Reply:
I am rounding in the java side. I wasn't sure whether I should write out progress as an integer (percentage) already from c++ or whether it is more flexible to let java decide what to do with it.

Reply:
I don't think it matters too much. The one small advantage I see in doing it all on the C++ side is you can tie the rounding here to the granularity at which you report above so you always get non-zero progress for each progress result written.

If we reckon reporting every 1% is suitable (seems reasonable to me) then as you say you rounding to an integer in the output seems like a nice choice.

Reply:
True, that would minimise the overhead. Perhaps that's simpler and makes more sense for now. We can always change it later. I'll do that.

Reply:
We have a precedent for reporting progress.  It's in `CForecastDataSink::writeStats`, and it writes progress as a proportion rather than a percentage.  So I think that for consistency it would be best if the `progress` field in the JSON doc is still a proportion rather than a percentage.  But there's definitely no need to report increments of less than 1%.